### PR TITLE
Introduce support for pluggable rego custom built-in functions in go

### PR DESCRIPTION
**Pull request by tareqmamari**
This commit introduces support for having pluggable rego custom built-in
functions in Go, this allows adapters to extend OPA in-process with
custom authorization logic.

Signed-off-by: Neyyan Sony <neyysony@in.ibm.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -23,6 +23,8 @@ require (
 	github.com/metalmatze/signal v0.0.0-20210307161603-1c9aa721a97a
 	github.com/mitchellh/mapstructure v1.4.1
 	github.com/oklog/run v1.1.0
+	github.com/onsi/ginkgo v1.16.3 // indirect
+	github.com/onsi/gomega v1.13.0 // indirect
```
Comment (matej-g) [inline go.mod:27]: Where is this coming from? :thinking:

File: opa/opa.go [hunk 2]
```diff
@@ -299,7 +323,17 @@ func NewInProcessAuthorizer(query string, paths []string, opts ...Option) (rbac.
 		o(c)
 	}
 
-	r := rego.New(rego.Query(query), rego.Load(paths, nil))
+	var r *rego.Rego
+
+	regoArgs := make([]func(*rego.Rego), 0, len(regoFunctions)+2)
```
Comment (matej-g) [inline opa/opa.go:328]: Why `+2`?

**Review by matej-g — COMMENTED** [review]
Looks good :raised_hands: Few smallish questions

File: go.mod (hunk 1 above)
Comment (tareqmamari) [reply go.mod:27]: This is indirect dependency from a kube package used by the Openshift authentication.

File: opa/opa.go (hunk 2 above)
Comment (tareqmamari) [reply opa/opa.go:328]: @matej-g `# of custom functions` + 2 ( the query and loading the paths args ), the reason why we moved out from dynamic slice is to make the lint happy.

Comment (matej-g) [reply opa/opa.go:328]: Ah, you're right, now I see it below.

**Comment by tareqmamari** [comment]
@observatorium-maintainers could you please look at this PR?

File: opa/opa.go [hunk 3]
```diff
@@ -24,6 +25,27 @@ const (
...
+//nolint:gochecknoglobals
+// regoFunctionsMtx is used to protect the providerFactories.
+var regoFunctionsMtx sync.RWMutex
+
+//nolint:gochecknoinits
+func init() {
+	regoFunctions = make(map[string]func(log.Logger) func(*rego.Rego))
+}
+
+// onboardNewFunction is used by the ...
```
Comment (matej-g) [inline opa/opa.go:41]: Unfinished comment perhaps?

**Review by matej-g — DISMISSED** [review]
Nicely done :star2: One more nit, otherwise looks good to me!

File: opa/opa.go (hunk 3 above)
Comment (tareqmamari) [reply opa/opa.go:41]: Ooops, that is a leftover from the initial PoC, I will update it.

**Review by matej-g — APPROVED** [review]
All looking good now, thanks @tareqmamari for your work!

**Pull request merged**